File: code/Java/src/main/java/Infrastructure/Bus/ServiceBus.java
package Infrastructure.Bus;

import Infrastructure.IProcessMessages;
import Infrastructure.IReadMessages;
import Infrastructure.ISendNotifications;

public class ServiceBus implements
        IMessageReceivedBus,
        ISendFinalStateBus,
        ISendNotificationBus,
        IReadMessages {
    private IProcessMessages marsRoverController;
    private ISendNotifications marsRoverSender;

    public void callBack(IProcessMessages marsRoverController) {
        this.marsRoverController = marsRoverController;
    }

    public void trigger(ISendNotifications marsRoverSender) {
        this.marsRoverSender = marsRoverSender;
    }

    public void NotifyMessageReceived(String rebuiltMessage) {
        this.marsRoverController.process(rebuiltMessage);
    }

    public void NotifyExecution(String finalState) {
        this.marsRoverSender.send(finalState);
    }
}

File: code/CSharp/ConnascentMarsRover/Source/Infrastructure/Bus/ServiceBus.cs
namespace Source.Infrastructure.Bus
{
    public class ServiceBus : 
        IMessageReceivedBus,
        ISendFinalStateBus,
        ISendNotificationBus,
        IReadMessages
    {
        private IProcessMessages marsRoverController;
        private ISendNotifications marsRoverSender;

        public void CallBack(IProcessMessages marsRoverController)
        {
            this.marsRoverController = marsRoverController;
        }

        public void Trigger(ISendNotifications marsRoverSender)
        {
            this.marsRoverSender = marsRoverSender;
        }

        public void NotifyMessageReceived(string rebuiltMessage)
        {
            this.marsRoverController.Process(rebuiltMessage);
        }

        public void NotifyError()
        {
            this.marsRoverSender.SendError();
        }

        public void NotifyExecution(string finalState)
        {
            this.marsRoverSender.Send(finalState);
        }
    }
}